### PR TITLE
[Frontend][PyTorch] Add: Relay stft operator

### PR DESCRIPTION
**Pull request by jsheng-jian**
This PR adds the stft, amax, amin 
[torch.stft](https://pytorch.org/docs/stable/generated/torch.stft.html)
```  
# This is run using 4 cpu cores for 100 iterations average inference time per frame
pt Elapsed time: 0.1826983118057251
graph_runtime Elapsed time: 0.11588997602462768
```

### CODE REVIEW

**Comment by masahi** [comment]
@jsheng-jian Thanks, looks like `topi/stft.py` and `topi/cuda/stft.py` are missing?

I'm very curious how you implemented stft (without fft)!

File: python/setup.py [hunk 1]
```diff
@@ -87,7 +87,9 @@ def git_describe_version(original_version):
 
 
 LIB_LIST, __version__ = get_lib_path()
+print("%%%%%%%%%%%%%%%%%%%%%%%: ", __version__)
 __version__ = git_describe_version(__version__)
+print("#######################: ", __version__)
```
Comment (masahi) [inline python/setup.py:92]: Remove

Comment (jsheng-jian) [reply python/setup.py]: Hi @masahi , thank you for looking at this! I have removed this.

**Comment by jsheng-jian** [comment]
`topi/stft.py` and `topi/cuda/stft.py` are added, could you take another look? @masahi

**Comment by masahi** [comment]
Please address the CI issue, there is a warning error from doc.

File: python/tvm/relay/op/transform.py [hunk 2]
```diff
@@ -1823,3 +1823,40 @@ def invert_permutation(data):
...
+    data : relay.Expr
+        Either a 1-D tensor or a 2-D batch tensor.
+    n_fft : int
+        The size of Fourier transform
+    hop_length : int
+        The distance between neighboring sliding window frames
+    win_length : int
+        The size of window frame and STFT filter
+    window : relay.Expr
+        A 1-D tensor window frame
```
Comment (masahi) [inline python/tvm/relay/op/transform.py:1843]: In PyTorch, window argument is optional. So shouldn't we support that too?

File: python/tvm/relay/op/transform.py [hunk 3]
```diff
@@ -1823,3 +1823,40 @@ def invert_permutation(data):
...
+    normalized : bool
+        Whether to return the normalized STFT results
+    onesided : bool
+        Whether to return onesided result or fill with conjugate symmetry
+    Returns
+    -------
+    output : relay.Expr
+        Tensor containing the STFT result
```
Comment (masahi) [inline python/tvm/relay/op/transform.py:1851]: Document output shape

File: python/tvm/relay/op/transform.py (hunk 2 above)
Comment (masahi) [inline python/tvm/relay/op/transform.py:1843]: In PyTorch, window argument is optional. So shouldn't we support that too?

File: python/tvm/relay/op/transform.py (hunk 3 above)
Comment (masahi) [inline python/tvm/relay/op/transform.py:1851]: Document the output shape. I had to read the type rel to see how the output shape looks like.

File: python/tvm/topi/cuda/stft.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            nthread_bx = ceil_div(output_ptr.shape[0], max_threads)
+            tx = te.thread_axis("threadIdx.x")
+            bx = te.thread_axis("blockIdx.x")
+            ib.scope_attr(tx, "thread_extent", nthread_tx)
+            ib.scope_attr(bx, "thread_extent", nthread_bx)
+            batch = bx * max_threads + tx
+
+            with ib.for_range(0, output_ptr.shape[0]) as batch:
+                with ib.for_range(0, output_ptr.shape[1]) as row:
+                    with ib.for_range(0, output_ptr.shape[2]) as col:
```
Comment (masahi) [inline python/tvm/topi/cuda/stft.py:96]: This looks weird. You try to parallelize over the batch dim but nothing is parallelized.

File: python/tvm/topi/stft.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        loop_kind,
+    ):
+        ib = tir.ir_builder.create()
+        data = ib.buffer_ptr(data_ptr)
+        window = ib.buffer_ptr(window_ptr)
+        output = ib.buffer_ptr(output_ptr)
+
+        with ib.for_range(0, output_ptr.shape[0]) as batch:
+            # https://librosa.org/doc/0.7.2/_modules/librosa/core/spectrum.html#stft
+            with ib.for_range(0, output_ptr.shape[1], kind="parallel") as row:
```
Comment (masahi) [inline python/tvm/topi/stft.py:85]: fuse the outer loop to have one big parallel loop.

File: python/tvm/topi/stft.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                        )
+                    with ib.if_scope(normalized):
+                        output[batch, row, col, 0] /= tir.sqrt(tir.const(n_fft, "float32"))
+                        output[batch, row, col, 1] /= tir.sqrt(tir.const(n_fft, "float32"))
+
+        return ib.get()
+
+    output_buf = tir.decl_buffer(output_shape, data.dtype, "output_buf")
+    loop_kind = "vectorize"
+    if hasattr(output_shape[2], "name") and output_shape[2].name == "any_dim":
```
Comment (masahi) [inline python/tvm/topi/stft.py:108]: `if isinstance(output_shape[2], tir.expr.Any)`

File: tests/python/frontend/pytorch/test_forward.py [hunk 7]
```diff
@@ -4113,6 +4125,36 @@ def test_fn(equation):
...
+    verify_trace_model(test_fn(3, 3, 3, False, "reflect", False, True), [input, window], targets)
+    verify_trace_model(test_fn(3, 3, 3, True, "reflect", False, True), [input, window], targets)
+    verify_trace_model(test_fn(3, 3, 3, True, "reflect", True, True), [input, window], targets)
+    verify_trace_model(test_fn(3, 3, 3, True, "reflect", False, False), [input, window], targets)
+    input = torch.rand([2, 12]).float()
+    window = torch.tensor([2, 3, 4], dtype=torch.int32)
+    verify_trace_model(test_fn(3, 3, 3, False, "reflect", False, True), [input, window], targets)
+    window = torch.tensor([1, 3], dtype=torch.int32)
+    verify_trace_model(test_fn(2, 1, 2, False, "reflect", False, True), [input, window], targets)
+
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_forward.py:4156]: Please add a test for `window=None` case.

File: tests/python/relay/test_op_level3.py [hunk 8]
```diff
@@ -1784,23 +1784,6 @@ def test_segment_sum(
...
-                ref_res
-            ), "Outputs from TVM and Python implementation must be equal "
-
-            for op_result, ref_result in zip(op_res, ref_res):
-                tvm.testing.assert_allclose(op_result.numpy(), ref_result, rtol=1e-5)
-        else:
-            tvm.testing.assert_allclose(op_res.numpy(), ref_res, rtol=1e-5)
-        relay.backend.te_compiler.get().clear()
-
-
```
Comment (masahi) [inline tests/python/relay/test_op_level3.py:1803]: Remove this diff

File: python/tvm/relay/op/transform.py [hunk 9]
```diff
@@ -1823,3 +1823,40 @@ def invert_permutation(data):
         relay.invert_permutation(data) = [2, 4, 3, 0, 1]
     """
     return _make.invert_permutation(data)
+
+
+def stft(data, n_fft, hop_length, win_length, window, normalized, onesided):
+    """
+    The STFT computes the Fourier transform of short overlapping windows of the input.
+    This giving frequency components of the signal as they change over time.
```
Comment (masahi) [inline python/tvm/relay/op/transform.py:1831]: giving -> gives

Fix the same typo in other files too.

**Review by masahi — CHANGES_REQUESTED** [review]

File: tests/python/relay/test_op_level3.py [hunk 10]
```diff
@@ -2066,5 +2066,170 @@ def verify_unique(n, dtype, is_dyn=False, is_sorted=False, return_counts=False):
...
+            backends = ["graph", "vm"]
+
+        z = relay.op.stft(data, n_fft, hop_length, win_length, window, normalized, onesided)
+        func = relay.Function([data, window], z)
+        verify_func2(
+            target, dev, func, [data_np, window_np], ref_res, rtol=1e-3, atol=1e-3, kinds=backends
+        )
+
+
+def verify_func2(target, dev, func, data, ref_res, rtol=1e-5, atol=1e-7, kinds=["vm"]):
```
Comment (masahi) [inline tests/python/relay/test_op_level3.py:2218]: Why do you need this? It looks identical to `verify_func`.

File: python/tvm/topi/cuda/stft.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def stft(
+    data,
+    n_fft,
+    hop_length,
+    win_length,
+    window,
+    normalized,
+    onesided,  # pylint: disable=unused-argument
```
Comment (masahi) [inline python/tvm/topi/cuda/stft.py:32]: You can remove `# pylint: disable=unused-argument` and add `unused-argument` to L17.

File: python/tvm/topi/cuda/stft.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    """
+
+    def gen_ir(
+        data_ptr,
+        n_fft,
+        hop_length,
+        win_length,
+        window_ptr,
+        normalized,
+        onesided,  # pylint: disable=unused-argument
```
Comment (masahi) [inline python/tvm/topi/cuda/stft.py:76]: same

File: python/tvm/topi/stft.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+
+def stft(
+    data,
+    n_fft,
+    hop_length,
+    win_length,
+    window,
+    normalized,
+    onesided,  # pylint: disable=unused-argument
```
Comment (masahi) [inline python/tvm/topi/stft.py:30]: same as the comment in `topi/cuda/stft.py`

File: python/tvm/topi/stft.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    """
+
+    def gen_ir(
+        data_ptr,
+        n_fft,
+        hop_length,
+        win_length,
+        window_ptr,
+        normalized,
+        onesided,  # pylint: disable=unused-argument
```
Comment (masahi) [inline python/tvm/topi/stft.py:74]: same

File: python/tvm/relay/op/transform.py (hunk 2 above)
Comment (jsheng-jian) [reply python/tvm/relay/op/transform.py]: Added optional window argument.

File: python/tvm/relay/op/transform.py (hunk 3 above)
Comment (jsheng-jian) [reply python/tvm/relay/op/transform.py]: Updated.

File: python/tvm/topi/cuda/stft.py (hunk 4 above)
Comment (jsheng-jian) [reply python/tvm/topi/cuda/stft.py]: Updated Cuda implementation.

File: python/tvm/topi/stft.py (hunk 5 above)
Comment (jsheng-jian) [reply python/tvm/topi/stft.py]: Done.

File: python/tvm/topi/stft.py (hunk 6 above)
Comment (jsheng-jian) [reply python/tvm/topi/stft.py]: The type is `tir.expr.SizeVar`, updated.

File: tests/python/frontend/pytorch/test_forward.py (hunk 7 above)
Comment (jsheng-jian) [reply tests/python/frontend/pytorch/test_forward.py:4157]: Added

File: tests/python/relay/test_op_level3.py (hunk 8 above)
Comment (jsheng-jian) [reply tests/python/relay/test_op_level3.py]: Done

File: python/tvm/relay/op/transform.py (hunk 9 above)
Comment (jsheng-jian) [reply python/tvm/relay/op/transform.py]: Fixed.

File: tests/python/relay/test_op_level3.py (hunk 10 above)
Comment (jsheng-jian) [reply tests/python/relay/test_op_level3.py:2218]: I want to expost `rtol` `atol` and `kinds`, I can update the original function instead of creating a new one?

File: python/tvm/topi/cuda/stft.py (hunk 11 above)
Comment (jsheng-jian) [reply python/tvm/topi/cuda/stft.py]: Done.

File: python/tvm/topi/cuda/stft.py (hunk 12 above)
Comment (jsheng-jian) [reply python/tvm/topi/cuda/stft.py]: Done.

File: python/tvm/topi/stft.py (hunk 13 above)
Comment (jsheng-jian) [reply python/tvm/topi/stft.py]: Done.

File: python/tvm/topi/stft.py (hunk 14 above)
Comment (jsheng-jian) [reply python/tvm/topi/stft.py]: Done.

File: tests/python/relay/test_op_level3.py (hunk 10 above)
Comment (masahi) [reply tests/python/relay/test_op_level3.py:2218]: Yeah that would be better

**Review by masahi — APPROVED** [review]

**Comment by jsheng-jian** [comment]
The CI failure seems to be unrelated.

**Comment by jsheng-jian** [comment]
It looks the CI failure is resolved in main branch, do I need to rebase my changes?

**Comment by masahi** [comment]
Yes, please send another job
